### PR TITLE
Make select(2) return with EINTR on signal

### PR DESCRIPTION
**Pull request by mame**
If a signal handler is invoked while doing `select(2)`, it must return with `EINTR`.

```c
#include <stdio.h>
#include <unistd.h>
#include <signal.h>
#include <sys/select.h>
#include <errno.h>

void handle_sigint(int n) {
  puts("SIGINT");
}

int main() {
  signal(SIGINT, handle_sigint);
  for(;;) {
    fd_set rfds;
    FD_ZERO(&rfds);
    FD_SET(0, &rfds);

    struct timeval tv;
    tv.tv_sec = 5;
    tv.tv_usec = 0;

    puts("wait for 5 seconds by select(2)");
    int ret = select(1, &rfds, NULL, NULL, &tv);
    printf("ret: %d, errno: %d\n", ret, errno);
    if (ret > 0) {
      char c;
      read(0, &c, 1);
      printf("read: %d\n", c);
    }
  }
}
```

Expected:
```
wait for 5 seconds by select(2)
^CSIGINT
ret: -1, errno: 4
wait for 5 seconds by select(2)
^CSIGINT
ret: -1, errno: 4
```

Actual:
```
wait for 5 seconds by select(2)
SIGINT
^CSIGINT
^CSIGINT
^CSIGINT
^CSIGINT
^CSIGINT
^Cret: 0, errno: 0
wait for 5 seconds by select(2)
SIGINT
^CSIGINT
^CSIGINT
^Cret: 0, errno: 0
```

Background: I supported xterm resizing in #27. Resize events are notified by SIGWINCH, but since select does not return, the screen of vim is not updated until input.

@RReverser If you have any comments I'd be glad to hear them.

### CODE REVIEW

File: emscripten-pty.js [hunk 1]
```diff
@@ -168,7 +176,20 @@ Object.assign(Lib, {
         // If so, that means it called into the PTY and the buffer was empty.
         if (result === -{{{ 1000 + cDefs.EAGAIN }}}) {
             // Wait for the PTY to become readable and try again.
-            PTY_waitForReadable(ok => wakeUp(ok ? impl() : 0));
+            PTY_waitForReadable(type => {
+                console.log(type);
```
Comment (RReverser) [inline emscripten-pty.js:180]: Stray debug leftover?

File: emscripten-pty.js [hunk 2]
```diff
@@ -89,20 +89,28 @@ Object.assign(Lib, {
...
         }
-        let handler, timeoutId;
+        let handlerReadable, handlerSignal, timeoutId;
         new Promise((resolve) => {
-            handler = PTY.onReadable(() => resolve(true));
+            handlerReadable = PTY.onReadable(() => resolve(0 /* ready */));
+
+            // We need to stop select(2) when a signal is caught.
+            //
+            // TODO: In fact, it should be stopped "when a signal handler is called," not "a the signal is received."
```
Comment (RReverser) [inline emscripten-pty.js:98]: You should be able to use `sigaction` to check it.

File: emscripten-pty.js [hunk 3]
```diff
@@ -89,20 +89,28 @@ Object.assign(Lib, {
...
+            // If the signal handler is set as SIG_IGN, select(2) should not be stopped.
```
Comment (RReverser) [inline emscripten-pty.js:99]: You should be able to use `sigaction` to check it.

**Comment by RReverser** [comment]
> @RReverser If you have any comments I'd be glad to hear them.

Sorry, those edge cases are a bit beyond me. I'm happy to help with Emscripten / JS parts but less familiar with syscall stuff - e.g. I didn't even know `select` can stop on signals.

File: emscripten-pty.js (hunk 1 above)
Comment (mame) [reply emscripten-pty.js]: Ah, thanks! Fixed.

File: emscripten-pty.js (hunk 3 above)
Comment (mame) [reply emscripten-pty.js:99]: Can I use `sigaction` from the JS side? Could you give me any pointers or examples?

Comment (RReverser) [reply emscripten-pty.js:99]: Sure, look at how `raise` is used in the same file already.

You can use any C function by adding like `'sigaction' ` to the relevant `__deps` array and calling it as `_sigaction` (note the unfortunate historical difference in underscore prefix).

Comment (mame) [reply emscripten-pty.js:99]: Thanks, I got it working with a patch as follows.

```diff
diff --git a/emscripten-pty.js b/emscripten-pty.js
index d81d51a..c02624c 100644
--- a/emscripten-pty.js
+++ b/emscripten-pty.js
@@ -84,7 +84,7 @@ Object.assign(Lib, {
         throw new FS.ErrnoError({{{ 1000 + cDefs.EAGAIN }}});
     },

-    $PTY_waitForReadableWithCallback__deps: ['$PTY', '$PTY_pollTimeout'],
+    $PTY_waitForReadableWithCallback__deps: ['malloc', 'free', 'sigaction', '$PTY', '$PTY_pollTimeout'],
     $PTY_waitForReadableWithCallback: (callback) => {
         if (PTY_pollTimeout === 0) {
             return callback(PTY.readable);
@@ -97,7 +97,16 @@ Object.assign(Lib, {
             //
             // TODO: In fact, it should be stopped "when a signal handler is called," not "a the signal is received."
             // If the signal handler is set as SIG_IGN, select(2) should not be stopped.
-            handlerSignal = PTY.onSignal(() => resolve(1 /* interrupted */));
+            handlerSignal = PTY.onSignal((signalName) => {
+                let signalCode = PTY_signalNameToCode[signalName];
+                var buff = _malloc({{{ POINTER_SIZE * 5 }}}); /* sizeof(struct sigaction) */
+                _sigaction(signalCode, null, buff);
+                var sa_handler = {{{ makeGetValue('buff', 0 /* sa_handler offset */, 'i32') }}};
+                _free(buff);
+                if (sa_handler != -2 /* SIG_IGN */) {
+                    resolve(1 /* interrupted */);
+                }
+            });

             if (PTY_pollTimeout >= 0) {
                 // if negative timeout, don't stop early (in poll-like functions it means infinite wait),
```

However, it was full of immediate values due to the lack of relevant struct and macro constant definitions.
Also, it should be handled for all signals, not just those generated by PTY.

Considering these things, I think that this support should be done on the emscripten side, not on xterm-pty. So I'd like to postpone this ad-hoc patch until it is truly needed.

**Pull request merged**